Add aws-firehose-receiver to support collecting AWS CloudWatch metric(OpenTelemetry format) (#10300)
apache/skywalking: docs/en/setup/backend/aws-firehose-receiver.md

## Setup(S3 example)

1. Create CloudWatch metrics configuration for S3 (refer to [S3 CloudWatch metrics](https://docs.aws.amazon.com/AmazonS3/latest/userguide/configure-request-metrics-bucket.html)) 
2. Stream CloudWatch metrics to AWS Kinesis Data Firehose delivery stream by [CloudWatch metrics stream](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-setup-datalake.html)
3. Specify AWS Kinesis Data Firehose delivery stream HTTP Endpoint (refer to [Choose HTTP Endpoint for Your Destination](https://docs.aws.amazon.com/firehose/latest/dev/create-destination.html#create-destination-http))

Usually, the [AWS CloudWatch metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/working_with_metrics.html) process flow with OAP is as follows:
```
CloudWatch metrics with S3 -->  CloudWatch Metric Stream (OpenTelemetry formart) --> Kinesis Data Firehose Delivery Stream --> AWS Firehose receiver(OAP) --> OpenTelemetry receiver(OAP)
```

## Notice

1. Only OpenTelemetry format is supported (refer to [Metric streams output formats](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats.html))

> 2. Only HTTPS could be accepted, you could directly enable TSL and set the receiver to listen 443, or put the receiver behind a gateway with HTTPS (refer to [Amazon Kinesis Data Firehose Delivery Stream HTTP Endpoint Delivery Specifications](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html))

Reviewer: ```suggestion
2. Only HTTPS could be accepted, you could directly enable TLS and set the receiver to listen 443, or put the receiver behind a gateway with HTTPS (refer to [Amazon Kinesis Data Firehose Delivery Stream HTTP Endpoint Delivery Specifications](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html))
```

# AWS Firehose receiver

> AWS Firehose receiver listen `127.0.0.1:13800` by default, and provides an HTTP Endpoint `/aws/firehose/metrics` that follows [Amazon Kinesis Data Firehose Delivery Stream HTTP Endpoint Delivery Specifications](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html)

Reviewer 2: As a k8s oriented setup, we should listen on 0.0.0.0.

```suggestion
AWS Firehose receiver listens on `0.0.0.0:13800` by default, and provides an HTTP Endpoint `/aws/firehose/metrics` that follows [Amazon Kinesis Data Firehose Delivery Stream HTTP Endpoint Delivery Specifications](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html)
```

You could leverage the receiver to collect [AWS CloudWatch metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/working_with_metrics.html), and analysis it through [MAL](../../concepts-and-designs/mal.md) as the receiver bases on [OpenTelemetry receiver](./opentelemetry-receiver.md)

> ## Setup(S3 example)

Reviewer 2: We will have a more specific doc for s3, right?

Author: Yep, this doc introduces how the receiver works in common way,  S3 (or other AWS service) metrics doc should be provided in itself PR

1. Create CloudWatch metrics configuration for S3 (refer to [S3 CloudWatch metrics](https://docs.aws.amazon.com/AmazonS3/latest/userguide/configure-request-metrics-bucket.html)) 
2. Stream CloudWatch metrics to AWS Kinesis Data Firehose delivery stream by [CloudWatch metrics stream](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-setup-datalake.html)
3. Specify AWS Kinesis Data Firehose delivery stream HTTP Endpoint (refer to [Choose HTTP Endpoint for Your Destination](https://docs.aws.amazon.com/firehose/latest/dev/create-destination.html#create-destination-http))

Usually, the [AWS CloudWatch metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/working_with_metrics.html) process flow with OAP is as follows:
```
CloudWatch metrics with S3 -->  CloudWatch Metric Stream (OpenTelemetry formart) --> Kinesis Data Firehose Delivery Stream --> AWS Firehose receiver(OAP) --> OpenTelemetry receiver(OAP)
```

## Notice

1. Only OpenTelemetry format is supported (refer to [Metric streams output formats](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-metric-streams-formats.html))

> 2. Doesn't support HTTP Headers - Content-Encoding

Reviewer 2: ```suggestion
2. Don't support HTTP Headers - Content-Encoding
```

Reviewer 2: What does this mean? Where and how  could users setup this?

Author: Our HTTPServer doesn't support Content-Encoding such as gzip,  the issue could be resolved through `ServerBuilder#decorator(DecodingService.newDecorator())`, I will submit it later

Reviewer 2: OK, got it. Felt strange when read this.

> 3. Only HTTPS could be accepted, you could directly enable TSL and set the receiver to listen 443, or put the receiver behind a gateway with HTTPS (refer to [Amazon Kinesis Data Firehose Delivery Stream HTTP Endpoint Delivery Specifications](https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html))

Reviewer 2: Is 443 or 13800 expected to set? I found inconsistent in the doc

Author: The default port is `13800`. But if users want to directly use the receiver, they have to set the port to 443, because of require of AWS

Reviewer 2: If AWS requires this, we should set 443 by default. As the HTTP server is off by OAP, we should be good.

Author: 443 requires root privilege, It will force the user to run OAP by root. So I choose another port as default and introduce that the port could be modified in the doc.

Reviewer 2: Does AWS support use custom port? This only makes sense if aws did.

Reviewer: > Does AWS support use custom port? This only makes sense if aws did.

It makes sense when users run OAP behind a gateway. In my opinion it makes no sense to expose port 443 (https port) from OAP, it's hardly possible in reality. I'm still in doubt whether this is useful and practical in reality.

Reviewer 2: If we need a gateway, we should document it. But the reason is the key.

In this thread, two things we need to make sure.
- 443 is required or just default
- whether a gatway is recommended, and why.
